crates/gitbutler-repo/src/repository_ext.rs

```diff
@@ -156,73 +157,67 @@ impl RepositoryExt for git2::Repository {
-                        (id, gix::object::tree::EntryKind::Link)
-                    } else {
-                        let mut file = std::fs::File::open(&path).with_context(|| {
-                            format!(
-                                "Could not open file at '{}' for adding it to the object database",
-                                path.display()
-                            )
-                        })?;
-                        let kind = if gix::fs::is_executable(&md) {
-                            gix::object::tree::EntryKind::BlobExecutable
-                        } else {
-                            gix::object::tree::EntryKind::Blob
-                        };
-                        (repo.write_blob_stream(&mut file)?, kind)
-                    };
-
-                    head_tree_editor.upsert(rela_path, kind, id)?;
-                }
-                gix::status::index_worktree::iter::Item::Rewrite { .. } => {
-                    unreachable!("disabled")
-                }
-                _ => {}
+        let mut tree_update_builder = git2::build::TreeUpdateBuilder::new();
+
+        let worktree_path = self.worktree_path();
+
+        let statuses = self.statuses(Some(
+            StatusOptions::new()
+                .renames_from_rewrites(false)
+                .renames_head_to_index(false)
+                .renames_index_to_workdir(false)
+                .include_untracked(true)
+                .recurse_untracked_dirs(true),
+        ))?;
+
+        // Truth table for upsert/remove:
+        // | HEAD Tree -> Index | Index -> Worktree | Action    |
+        // | add                | delete            | no-action |
+        // | modify             | delete            | remove    |
+        // |                    | delete            | remove    |
+        // | delete             |                   | remove    |
+        // | delete             | add               | upsert    |
+        // | add                |                   | upsert    |
+        // |                    | add               | upsert    |
+        // | add                | modify            | upsert    |
+        // | modify             | modify            | upsert    |
+
+        for status_entry in &statuses {
+            let status = status_entry.status();
+            let path = status_entry.path().context("Failed to get path")?;
+            let path = Path::new(path);
+
+            if status.is_index_new() && status.is_wt_deleted() {
+                // This is a no-op
+            } else if (status.is_index_deleted() && !status.is_wt_new()) || status.is_wt_deleted() {
+                tree_update_builder.remove(path);
+            } else {
+                let file_path = worktree_path.join(path);
+                let file = std::fs::read(file_path)?;
+                let blob = self.blob(&file)?;
+
+                tree_update_builder.upsert(path, blob, git2::FileMode::Blob);
```

Comment:
There are various [file-modes](https://docs.rs/git2/latest/git2/enum.FileMode.html#) possible here and all but `Tree` and `Unreadable` (and probably, `BlobGroupWritable`) should be handled. 

`Link` is particularly important, `BlobExecutable` as well. These cases were handled in the deleted code, so could be brought back from there. Submodules also weren't handled though, maybe it would have failed, which seems OK since submodules aren't supported yet.